### ISSUE TITLE
Proposal to change POSIX EAGER interface to avoid extra packing buffer

### ISSUE THREAD

**Issue by yfguo** [pmodels/mpich#4794]
POSIX_am_isend today create a pack buffer for non-contiguous data and passes that buffer to POSIX_eager_send.

https://github.com/pmodels/mpich/blob/77199f67bfe2c3663e6c37476c9866479a6ef68c/src/mpid/ch4/shm/posix/posix_am.h#L130

For iqueue eager module, it will internally copies from the pack buffer into the cells. To me, this pack buffer is just an extra cost. This buffer should have been registered with GPU which make the cost even higher. I would like to remove it and directly packs into the cells of iqueue. I propose the following changes to make that happen.

1. Add a new interface `POSIX_eager_get_buffer` which returns the payload part of an iqueue cell and the size of it. POSIX will be packing data into it directly.
2. Modify the `POSIX_eager_send` interface to take that as a parameter and send it.

Note that this also mean the handling of segmented sending is moved from iqueue to POSIX which would be a substantial change in addition to the interface.

Alternatively, we could also make the EAGER interface to take datatype, count so it can handles non-contiguous datatype internally.

Looking for input from Intel about what is the best way to achieve the goal of removing the pack buffer.
@wesbland @ddurnov

**Comment by yfguo** [comment]
I have created https://github.com/pmodels/mpich/pull/4839 as an example of what I proposed.

I would also point out that the existing eager implementation has many interdependencies between POSIX and EAGER iqueue. Iqueue would update data structures created by POSIX, and it relies on a implicit "agreement" between POSIX and iqueue on how these data structures are handle.

What I suggest that iqueue (or an eager module) should do are 1) providing a SHM buffer that for carrying the data and 2) send that buffer to the target rank. It would be POSIX's responsibility to 1) make sure the data fits in the buffer provided by the eager module; 2) break large messages into smaller pieces and handle the pipelined send; 3) pack data if needed.
